hazelcast/src/main/java/com/hazelcast/sql/impl/expression/datetime/Formatter.java

```diff
@@ -0,0 +1,791 @@
+        }
+        if (formatter instanceof DateFormat && !(input instanceof Temporal)) {
+            throw QueryException.dataException("Input parameter is expected to be date/time");
+        }
+        if (formatter instanceof NumberFormat && !(input instanceof Number)) {
+            throw QueryException.dataException("Input parameter is expected to be numeric");
+        }
+        return formatter.apply(input, Locale.forLanguageTag(locale));
+    }
+
+    private static final int[] ARABIC = {1000,  900, 500,  400, 100,   90,  50,   40,  10,    9,   5,    4,   1};
+    private static final String[] ROMAN = {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};
+    private static final String[] ORDINAL = {"th", "st", "nd", "rd", "th", "th", "th", "th", "th", "th"};
+
+    private static StringBuilder toRoman(int number) {
+        StringBuilder r = new StringBuilder();
+        for (int j = 0; j < ARABIC.length; j++) {
+            for (; number >= ARABIC[j]; number -= ARABIC[j]) {
+                r.append(ROMAN[j]);
+            }
+        }
+        return r;
+    }
+
+    private static String getOrdinal(String number) {
+        return number.endsWith("11") || number.endsWith("12") || number.endsWith("13")
+                ? "th" : ORDINAL[number.charAt(number.length() - 1) - 48];
+    }
+
+    private static class DateFormat implements BiFunction<Object, Locale, String> {
+        private final List<Part> parts = new ArrayList<>();
+
+        interface Part {
+            void format(StringBuilder s, Temporal input, Locale locale);
+        }
+
+        static class Literal implements Part {
+            final String contents;
+
+            Literal(String contents) {
+                this.contents = contents;
+            }
+
+            @Override
+            public void format(StringBuilder s, Temporal input, Locale locale) {
+                s.append(contents);
+            }
+
+            @Override
+            public String toString() {
+                return '"' + contents + '"';
+            }
+        }
+
+        static class PatternInstance implements Part {
+            final boolean fill, translate;
+            final Pattern pattern;
+            final Ordinal ordinal;
+
+            PatternInstance(boolean fill, boolean translate, Pattern pattern, Ordinal ordinal) {
+                this.fill = fill;
+                this.translate = translate;
+                this.pattern = pattern;
+                this.ordinal = ordinal;
+            }
+
+            @Override
+            public void format(StringBuilder s, Temporal input, Locale locale) {
+                Object result = pattern.query.apply(input, translate ? locale : Locale.US);
+                String r = result.toString();
+                if (pattern == Pattern.TZH) {
+                    s.append((int) result < 0 ? '-' : '+');
+                    r = (int) result < 0 ? r.substring(1) : r;
+                }
+                if (fill) {
```

Comment:
This is inverted meaning of "fill mode", we should pad if FM is not enabled.

Reply:
`FM` is a flag to enable/disable the fill mode. See the fill mode's definition in the class Javadoc.

Reply:
See the fill mode's definition in PG. Or rather, reverse-enginner it, I think it's not defined in PG's manuals.